src/app/fdctl/run/run.c

```diff
@@ -99,50 +75,122 @@ int
-    .tile_name = args->tile->name,
-  };
 
-  FD_TEST( tile_args.wksp_pod );
-  FD_TEST( args->tile->allow_workspaces_cnt <= sizeof(tile_args.wksp_pod) );
-  for( ulong i=0; i<args->tile->allow_workspaces_cnt; i++ ) {
-    /* preload pods before sandboxing, so workspace memory is already mapped */
-    tile_args.wksp_pod[ i ] = workspace_pod_join( args->app_name, workspace_kind_str( args->tile->allow_workspaces[ i ] ) );
+  /* calling fd_tempo_tick_per_ns requires nanosleep, it is cached with
+     a FD_ONCE.  We do this for all tiles before sandboxing so that we
+     don't need to allow the nanosleep syscall. */
+  fd_tempo_tick_per_ns( NULL );
+
+  /* preload shared memory before sandboxing, so it is already mapped */
+  fd_topo_join_tile_workspaces( args->config->name, &args->config->topo, tile );
+
+  fd_tile_config_t * config = fd_topo_tile_to_config( tile );
+
+  void * scratch_mem   = NULL;
+  if( FD_LIKELY( config->scratch_align ) ) {
+    scratch_mem = (uchar*)args->config->topo.workspaces[ tile->wksp_id ].wksp + tile->user_mem_offset;
+    if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)scratch_mem, config->scratch_align() ) ) )
+      FD_LOG_ERR(( "scratch_mem is not aligned to %lu", config->scratch_align() ));
   }
 
-  if( FD_UNLIKELY( args->tile->init ) ) args->tile->init( &tile_args );
+  if( FD_UNLIKELY( config->privileged_init ) )
+    config->privileged_init( &args->config->topo, tile, scratch_mem );
 
   int allow_fds[ 32 ];
-  ulong allow_fds_sz = args->tile->allow_fds( &tile_args,
-                                              sizeof(allow_fds)/sizeof(allow_fds[0]),
-                                              allow_fds );
-
-  fd_sandbox( args->sandbox,
-              args->uid,
-              args->gid,
-              allow_fds_sz,
+  ulong allow_fds_cnt = config->allow_fds( scratch_mem,
+                                           sizeof(allow_fds)/sizeof(allow_fds[ 0 ]),
+                                           allow_fds );
+
+  fd_sandbox( args->config->development.sandbox,
+              args->config->uid,
+              args->config->gid,
+              allow_fds_cnt,
               allow_fds,
-              args->tile->allow_syscalls_cnt,
-              args->tile->allow_syscalls );
-  args->tile->run( &tile_args );
+              config->allow_syscalls_cnt,
+              config->allow_syscalls );
+
+  /* Now we are sandboxed, join all the tango IPC objects in the workspaces */
+  fd_topo_fill_tile( &args->config->topo, tile, FD_TOPO_FILL_MODE_JOIN );
+  FD_TEST( tile->cnc );
+
+  if( FD_UNLIKELY( config->unprivileged_init ) )
+    config->unprivileged_init( &args->config->topo, tile, scratch_mem );
+
+  const fd_frag_meta_t ** in_mcache = fd_alloca( alignof(fd_frag_meta_t *), sizeof(fd_frag_meta_t *)*args->tile->in_cnt );
+  ulong ** in_fseq = fd_alloca( alignof(ulong *), sizeof(ulong *)*args->tile->in_cnt );
+  if( FD_UNLIKELY( !in_mcache || !in_fseq ) ) FD_LOG_ERR(( "fd_alloca failed" ));
```

Comment:
These already have static maximums (right?), so you could get rid of the alloca here, but it's also fine as is